### ISSUE TITLE
Generating RestResponse subclasses

### ISSUE THREAD

**Issue by RikkiGibson** [Azure/autorest.java#153]
`Single<RestResponse<MyGetHeaders, Flowable<ByteBuffer>>> getWithRestResponseAsync()` has quite a daunting return type.
I think we should consider generating simply `Single<MyGetResponse> getAsync()` where MyGetResponse is:

```java
public final class MyGetResponse extends RestResponse<MyGetHeaders, Flowable<ByteBuffer>> {
    MyGetResponse(int statusCode, MyGetHeaders headers, Map<String, String> rawHeaders, Flowable<ByteBuffer> body) {
        super(...);
    }
}
```
It might be good to explicitly implement the members in the child class because when people go to definition on one of the inherited members and they only see TGeneric instead of the concrete type being used in context, it's going to be annoying and hard to follow what the types of things are in the code.

This should be done before we release a preview.

**Comment by JonathanGiles** [comment]
Generally (and naively, as I don't use these APIs) I would probably disagree with this proposal. Hiding the type doesn't really make things any simpler in my opinion - it just hides the complexity and requires the developer to do a little more work to understand. It might be useful to add some additional text that demonstrates how the API would be used before and after this proposed change.

**Comment by RikkiGibson** [comment]
Maybe so. I'm concerned about the very long variable declarations to hold return values being cumbersome, and about people's eyes glazing over when they see the return types in our libraries.

RestResponse is a container for status code, headers and body. Users won't know by glancing at a usage that the type arguments on the class refer to the headers and body types. The likely discovery process is that they'll get a handle on the return value and start dotting into it and seeing what properties are there.

An abridged `interface Blobs`:
```java
interface Blobs {
    /**
     * The Get Blob operation reads or downloads a blob from the system, including its metadata and properties. You can also call Get Blob to read a snapshot.
     *
     * @return the {@link Single&lt;RestResponse&lt;BlobGetHeaders, Flowable&lt;ByteBuffer&gt;&gt;&gt;} object if successful.
     */
    Single<RestResponse<BlobGetHeaders, Flowable<ByteBuffer>>> getWithRestResponseAsync(URL url);
}
```
An example usage:
```java
AsynchronousFileChannel out = ...
Blobs blobs = ...
Single<RestResponse<BlobGetHeaders, Flowable<ByteBuffer>>> blobSingle = blobs.getWithRestResponseAsync(url);
blobSingle
    .flatMapCompletable(res -> FlowableUtil.writeFile(res.body(), out))
    .doFinally(out::close)
    .subscribe(() -> System.out.println("done"));
```

Versus what I propose:

```java
interface Blobs {
    /**
     * The Get Blob operation reads or downloads a blob from the system, including its metadata and properties. You can also call Get Blob to read a snapshot.
     *
     * @return the {@link Single&lt;BlobsGetResponse&gt;} object if successful.
     */
    Single<BlobsGetResponse> getWithRestResponseAsync(URL url);
}
```
Usage:
```java
AsynchronousFileChannel out = ...
Blobs blobs = ...
Single<BlobsGetResponse> blobSingle = blobs.getWithRestResponseAsync(url);
blobSingle
    .flatMapCompletable(res -> FlowableUtil.writeFile(res.body(), out))
    .doFinally(out::close)
    .subscribe(() -> System.out.println("done"));
```

**Comment by JonathanGiles** [comment]
It's a compelling argument, but in effect I worry that you're trading method signature cleanliness for class noise. Do you have much insight into how many of these `MyGetResponse` type classes would be generated for a standard SDK? Is there a more comfortable middle ground where there can be a simpler set of generic types to reduce the number of additional API classes required?

**Comment by RikkiGibson** [comment]
Class noise is a meaningful concern. This would potentially be an extra class declared for every REST API method being exposed through the generator. The main reason for that is there's no way to define and reuse a "headers" type in a Swagger spec. Thus the generator already creates a distinct strongly typed "Headers" class to hold strongly typed response headers for each REST API method. A limited amount of "merging" of the Response classes might be possible in cases where no response headers are specified in Swagger (and therefore no Headers class gets generated).

From the import list of `Blobs.java` (don't mean to pick on Storage, but Blob Service is a good example where both response headers and bodies contain important information):

```
import com.microsoft.azure.storage.models.BlobAbortCopyHeaders;
import com.microsoft.azure.storage.models.BlobCopyHeaders;
import com.microsoft.azure.storage.models.BlobDeleteHeaders;
import com.microsoft.azure.storage.models.BlobGetHeaders;
import com.microsoft.azure.storage.models.BlobGetMetadataHeaders;
import com.microsoft.azure.storage.models.BlobGetPropertiesHeaders;
import com.microsoft.azure.storage.models.BlobLeaseHeaders;
import com.microsoft.azure.storage.models.BlobPutHeaders;
import com.microsoft.azure.storage.models.BlobSetBlobTierHeaders;
import com.microsoft.azure.storage.models.BlobSetMetadataHeaders;
import com.microsoft.azure.storage.models.BlobSetPropertiesHeaders;
import com.microsoft.azure.storage.models.BlobTakeSnapshotHeaders;
import com.microsoft.azure.storage.models.BlobUndeleteHeaders;
```

I think it would be hard to improve on this in the generator.

**Comment by JonathanGiles** [comment]
*JonathanGiles cries*

Man do I wish we could do something about this. Tell me if I'm side tracking the issue (we can discuss offline or elsewhere), but how much are these header APIs used? How do developers use them? I know I'll be unpopular for suggesting this, but could we get away with some degree of read-only Map approach for header information? :-)

**Comment by JonathanGiles** [comment]
Ha - I did a search and found what I was thinking of. Something along the lines of https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/http/HttpHeaders.html

**Comment by RikkiGibson** [comment]
The only purpose of these classes is to show users through the type system what headers are expected to be present on the response. We always include a Map<String, String> rawHeaders on the response as well.

We have an HttpHeaders type that we use internally but at some point seemed to consider an implementation detail of HttpClient and HttpPipeline and so on.

I think it’s fair to interrogate why the classes are there as they’re one of the factors we must multiply by in generation (such factors often seem to find their way in—consider the number of overloads generated per REST API method)

**Comment by JonathanGiles** [comment]
I am in the fortunate position where I can continue to be in the naive position and just keep asking dumb questions :-) If we can assume that there is a fixed set of header keys (even if it mutates as swagger files create new header keys, we can assume that for any given generation run it is fixed), could we not create a `HeaderKey` Enum of all header key names globally, and then put on the appropriate classes an `EnumSet<HeaderKey>` that represents the valid headers for that given response?

**Comment by RikkiGibson** [comment]
No, it's useful to have someone with an outside POV weigh in.

The static type information for response headers is useful. It enables us to automatically convert response headers to types like `List<Integer>` or `DateTime`, or use `@HeaderCollection` to gather the response headers with a shared prefix into a Map, which we use to support [`x-ms-meta`](https://docs.microsoft.com/en-us/rest/api/storageservices/get-blob-metadata#response-headers) in Blob Service.

A service-scoped enum for every header that's been defined on every response in the Swagger could have users resorting to inspecting the generated code usages of `EnumSet.of(...)` to find which headers are expected to be present on a particular response. There'd also be the question of type information-- Java doesn't support generic types in enums, so there'd have to be some way to make that information available to the runtime and to the user.

Going back to the current design of Headers classes--I think it's a question of whether the increased discoverability through the IDE is worth the number of classes. I think it is worth it even if it can be unsightly. I speculate that users are less likely to browse the package contents and get disoriented by a large number of classes in the `.models` subpackage than they are to check out some sample code and start playing around and dotting into values to find what they're looking for.

Some AutoRest generators even define a class per REST API method to contain all the parameters. This can reduce the pain in cases like Cognitive Services where there are 20 optional parameters, of which users usually only ever wish to specify values for a few. Obviously it's not that realistic to make the code generator ship pleasant-to-use APIs for every conceivable service design.

Reducing duplication of these sorts of model definitions might be good to consider on the AutoRest core level or even OpenAPI Specification level.

I think at this point with the amount of runway we have left for the 2.0 Java runtime/generator I'm going to try and move forward with a Response subtype per-method, with some kind of deduplication for when response definitions are shared as a nice-to-have.

**Comment by RikkiGibson** [comment]
Azure/autorest-clientruntime-for-java#393
Will open an autorest.java PR after we merge the runtime PR--this way we'll be able to publish thea runtime snapshot and get green CI for autorest.java

**Comment by RikkiGibson** [comment]
Work complete
